### ISSUE TITLE
Regex for triggerCharacters

### ISSUE THREAD

**Issue by tbodt** [Microsoft/language-server-protocol#459]
I'm writing a language server that uses Tern for javascript autocomplete, and I'd like to have it autocomplete my installed npm packages after `require('`. Tern supports this, but I have my triggerCharacters set to `['.']`, which means my server is not getting called for completions when I type `require('`. I could add `'` to the triggerCharacters, but that would mean autocomplete popups whenever I start typing a string, which is pretty useless since it just returns the names of variables in scope.

Would it be a good idea to support regexes for triggerCharacters? nvim completion manager does this, and it works pretty well with the regex `require\s*\(\s*['"][^)'"]*`. Then I guess it would have to be called something else since they aren't characters anymore.

**Comment by dbaeumer** [comment]
@jrieken I remember that you and @aeschli discussed this once. What was the outcome?

**Comment by jrieken** [comment]
In VS Code we won't do it - mostly because of performance considerations and because it leads to a false understanding. Providers must expect to always be called and having complex trigger-expression leads to a false sense of 'I am invoked only after such things'. Instead, we provide more context now, e.g. what was the trigger reason, iff what character triggered me etc. It's the responsibility of provider to quickly figure out that there is nothing to suggest from there side. 

Think of the `.` as trigger character and user typing the dot-character inside a comment. One might find a regex for that but it's easier to programmatically decide that. And after all, a user can always press ctrl+space after a dot inside a comment.

**Comment by dbaeumer** [comment]
@jrieken thanks for the explanation.

@tbodt the additional context @jrieken was pointing to is available in the LSP as well. It is availble here: https://github.com/Microsoft/vscode-languageserver-node/blob/master/protocol/src/protocol.ts#L1358

Any objection to close the issue?

**Comment by tbodt** [comment]
Sure, we can close.
